lib/experimental/colortext.nim

```diff
@@ -0,0 +1,765 @@
+import std/[terminal, unicode, sequtils, strformat, strutils]
+
+type
+  TermColor8Bit* = distinct uint8
+  TermColor8BitBg* = distinct TermColor8Bit
+  TermColor8BitFg* = distinct TermColor8Bit
+
+
+func col8*(color: uint8): TermColor8Bit = TermColor8Bit(color)
+func bg*(color: TermColor8Bit): TermColor8BitBg = TermColor8BitBg(color)
+func fg*(color: TermColor8Bit): TermColor8BitFg = TermColor8BitFg(color)
+
+func `==`*(a, b: TermColor8Bit): bool = a.uint8 == b.uint8
+
+
+type
+  ColStyle* = object
+    case use8Bit*: bool
```

Comment:
so you store per rune a boolean that says if it uses 8 bit or not? Seems very inefficient if you ask me.

Reply:
It uses 8-bit colors styling, not 8 bits per color

Reply:
> > I've added colored to Nim as well
> 
> You mean this one, correct? [nim-lang/Nim@20a21aa](https://github.com/nim-lang/Nim/commit/20a21aa1848726a60493432b7337ecbfd491f7ac)

Well define what you see as "correct". It is correct in terms of ANSI color escapes, and it is very tiny. But it is certainly not correct in terms of platform agnostic. It just supports one single (standardized) type of text coloring in a minimalistic way.

> > Then it seems to oversee many attributes of Unicode and how styling text is done in general,
> 
> I don't understand what attributes of Unicode I'm missing, or what I'm missing from text styling. I do support all 16-color and 256-color schemes, as well as all regular text styling attributes.

https://www.compart.com/en/unicode/U+200F
grapheme clusters will break vertical alignment as far as I can see


> That suffers from exactly the same issue as I've described in the PR title - all code is littered with logical checks, you can't remove colors after they have been added, you can't do anything else on the text after it has been colorized (alignment does not work properly, you can't pad things left/right, get length etc., only append new text)

Well except from the logical checks, that are cetainly bad/annoying. This is designed to be minimal, yes. Only implement what is actually necessary.

Maybe you should start with a spec/requirements sheet first, so that I can judge according to the spec.
 
> Again, this is not a high-performance string handling library, it is a create-and-write-out solution. I don't plan to store these strings anywhere, or try to constantly rearrange them beyond what is minimally necessary to produce formatted output. I can try and squeeze a couple of bytes from the color handling, for example by using 64-bit integer and some bit hacking if you want this, but considering it is almost guaranteed that I will call `stdout.write` or `stderr.write` afterwards for each constructed string, I see little to no reason to do these micro-optimizations and the do IO that would dwarf all the runtime.

Actually I did measurements in the past, string handling and measurements are a performance bottleneck in the compiler.

Reply:
> string handling and measurements are a performance bottleneck in the compiler.

Which part of the string handling - error reporting or PSym handing? Those are two different use cases, and this library is not going to be used to represent all existing string handling operations from now on, so general measurements hardly apply here. Yes, it is slower than concatenating `\e[31m` from const, but whether this becomes a bottleneck is another question.

Reply:
>  grapheme clusters will break vertical alignment as far as I can see

Is that common issue that comes up in handing of the compiler messages (which are mostly English text) and some basic Unicode drawing characters?

Reply:
> This is designed to be minimal, yes. Only implement what is actually necessary.

Yes, it is minimal and does exactly what was necessary at the time, which is to colorize couple string elements with hardcoded styles. That's where things usually start, and if there is no demand to change anything it is generally fine. But in this specific case I'm trying to add a module that would make it easier to quickly progress and iterate through different error message formats and styles, without getting constantly bogged down in details. 

How about we add red underline for the error message location, like clang/rust/elm etc. do? Or implement inline type diffs like clang does on templates? Maybe some other way of error formatting - without being able to even align text blocks to each other we won't get anywhere if we try to cling to the "minimal" approach here.

Reply:
> Is that common issue that comes up in handing of the compiler messages (which are mostly English text) and some basic Unicode drawing characters?
> 
> To be honest, I'm not even sure if I understand the issue at hand, I would need to read Unicode spec/documentation or something of that nature.

The question is, is this for compiler internal use only, or is it supposed to be a string coloring library exposed for the public to be used? If it is compiler internal only, I think you would get almost away with ASCII only, When it is a public API, you have to dig into the Unicode spec/documentation, there is no way to avoid its complexity.

Reply:
*For now* it is most for compiler use, just like `experimental/diff`, in the future it can be expanded as necessary and moved out of experimental